hnn_core/network.py

```diff
@@ -311,55 +322,75 @@ def _create_all_src(self):
+                Cell = type2class[src_type]
+                if src_type in ('L2_pyramidal', 'L5_pyramidal'):
+                    self.cells.append(Cell(gid, src_pos, self.params))
                 else:
-                    print("None of these types in Net()")
-                    exit()
+                    self.cells.append(Cell(gid, src_pos))
+
+                pc.cell(gid, self.cells[-1].connect_to_target(
+                        None, self.params['threshold']))
+
+            # external inputs are special types of pseudo-cells
+            # 'common': all cells impacted with identical TIMING of spike
+            # events. NB: cell types can still have different weights for how
+            # such 'common' spikes influence them
+            elif src_type == 'common':
+                # print('cell_type',cell_type)
+                # to find param index, take difference between REAL gid
+                # here and gid start point of the items
+                p_ind = gid - self.gid_dict['common'][0]
+
+                # new ExtFeed: target cell type irrelevant (None) since input
+                # timing will be identical for all cells
+                # XXX common_feeds is a list of dict (comp. unique_feeds)
+                self.common_feeds.append(
+                    ExtFeed(feed_type=src_type,
+                            target_cell_type=None,
+                            params=self.p_common[p_ind],
+                            gid=gid))
+
+                # now use the param index in the params and create
+                # the cell and artificial NetCon
+                pc.cell(gid, self.common_feeds[-1].connect_to_target(
+                        self.params['threshold']))
+
+            # external inputs can also be Poisson- or Gaussian-
+            # distributed, or 'evoked' inputs (proximal or distal)
+            # these are cell-specific ('unique')
+            elif src_type in self.p_unique.keys():
+                gid_target = gid - self.gid_dict[src_type][0]
+                target_cell_type = self.gid_to_type(gid_target)
+
+                # new ExtFeed, where now both feed type and target cell type
+                # specified because these feeds have cell-specific parameters
+                # XXX unique_feeds is a dict of dict (comp. common_feeds)
+                self.unique_feeds[src_type].append(
+                    ExtFeed(feed_type=src_type,
+                            target_cell_type=target_cell_type,
+                            params=self.p_unique[src_type],
+                            gid=gid))
+                pc.cell(gid,
+                        self.unique_feeds[src_type][-1].connect_to_target(
+                            self.params['threshold']))
             else:
-                print("None of these types in Net()")
-                exit()
+                raise ValueError('No parameters specified for external feed '
+                                 'type: %s' % src_type)
```

Comment:
Excellent!

We should add a test too, but I'm fine leaving this for another PR. The diff is pretty big already.